x-pack/legacy/plugins/lens/server/routes/existing_fields.ts

```diff
@@ -82,6 +86,166 @@ export async function existingFieldsRoute(setup: CoreSetup) {
+  context,
+  indexPatternId,
+  fromDate,
+  toDate,
+  timeFieldName,
+}: {
+  indexPatternId: string;
+  context: RequestHandlerContext;
+  fromDate?: string;
+  toDate?: string;
+  timeFieldName?: string;
+}) {
+  const {
+    indexPattern,
+    indexPatternTitle,
+    mappings,
+    fieldDescriptors,
+  } = await fetchIndexPatternDefinition(indexPatternId, context);
+
+  const fields = buildFieldList(indexPattern, mappings, fieldDescriptors);
+
+  const docs = await fetchIndexPatternStats({
+    fromDate,
+    toDate,
+    client: context.core.elasticsearch.dataClient,
+    index: indexPatternTitle,
+    timeFieldName: timeFieldName || indexPattern.attributes.timeFieldName,
+    fields,
+  });
+
+  return {
+    indexPatternTitle,
+    existingFieldNames: existingFields(docs, fields),
+  };
+}
+
+async function fetchIndexPatternDefinition(indexPatternId: string, context: RequestHandlerContext) {
+  const savedObjectsClient = context.core.savedObjects.client;
+  const requestClient = context.core.elasticsearch.dataClient;
+  const indexPattern = await savedObjectsClient.get('index-pattern', indexPatternId);
+  const indexPatternTitle = indexPattern.attributes.title;
+  // TODO: maybe don't use IndexPatternsFetcher at all, since we're only using it
+  // to look up field values in the resulting documents. We can accomplish the same
+  // using the mappings which we're also fetching here.
+  const indexPatternsFetcher = new IndexPatternsFetcher(requestClient.callAsCurrentUser);
+  const [mappings, fieldDescriptors] = await Promise.all([
+    requestClient.callAsCurrentUser('indices.getMapping', {
+      index: indexPatternTitle,
+    }),
+
+    indexPatternsFetcher.getFieldsForWildcard({
+      pattern: indexPatternTitle,
+      // TODO: Pull this from kibana advanced settings
+      metaFields,
+    }),
+  ]);
+
+  return {
+    indexPattern,
+    indexPatternTitle,
+    mappings,
+    fieldDescriptors,
+  };
+}
+
+/**
+ * Exported only for unit tests.
+ */
+export function buildFieldList(
+  indexPattern: SavedObject,
+  mappings: MappingResult,
+  fieldDescriptors: FieldDescriptor[]
+): Field[] {
+  const aliasMap = Object.entries(Object.values(mappings)[0].mappings.properties)
+    .map(([name, v]) => ({ ...v, name }))
+    .filter(f => f.type === 'alias')
+    .reduce((acc, f) => {
+      acc[f.name] = f.path;
+      return acc;
+    }, {} as Record<string, string>);
+
+  const descriptorMap = fieldDescriptors.reduce((acc, f) => {
+    acc[f.name] = f;
+    return acc;
+  }, {} as Record<string, FieldDescriptor>);
+
+  return JSON.parse(indexPattern.attributes.fields).map(
+    (field: { name: string; lang: string; scripted?: boolean; script?: string }) => {
+      const path =
+        aliasMap[field.name] || descriptorMap[field.name]?.subType?.multi?.parent || field.name;
+      return {
+        name: field.name,
+        isScript: !!field.scripted,
+        isAlias: !!aliasMap[field.name],
+        path: path.split('.'),
```

Comment:
I think you're allowed to have dots in field names without being nested, but it works fine for most use.